### PR TITLE
feat: log format option for wait and init containers. Fixes #8986

### LINKED ISSUE #8986: Log in JSON when monitoring a child workflow
# Summary

I use the "workflow of workflows" pattern, basically this: https://argoproj.github.io/argo-workflows/workflow-of-workflows/ . In doing so, when the parent workflow runs, it polls the status of the child workflow to see if it's running or if terminated successfully or not. When it does that, it logs like this:

```text 
17:45:29.261 main time="2022-06-16T17:45:29.261Z" level=info msg="failure condition '{status.phase in [Error Failed]}' evaluated false"
17:45:29.261 main time="2022-06-16T17:45:29.261Z" level=info msg="success condition '{status.phase == [Succeeded]}' evaluated false"
17:45:29.261 main time="2022-06-16T17:45:29.261Z" level=info msg="0/1 success conditions matched"
17:45:29.261 main time="2022-06-16T17:45:29.261Z" level=info msg="Waiting for resource workflow.argoproj.io/dev-workflow-z8tr4 in namespace default resulted in retryable error: Neither success condition nor the failure condition has been matched. Retrying..."
17:45:34.260 main time="2022-06-16T17:45:34.260Z" level=info msg="Get workflows 200"
17:45:34.261 main time="2022-06-16T17:45:34.261Z" level=info msg="failure condition '{status.phase in [Error Failed]}' evaluated false"
17:45:34.261 main time="2022-06-16T17:45:34.261Z" level=info msg="success condition '{status.phase == [Succeeded]}' evaluated false"
17:45:34.261 main time="2022-06-16T17:45:34.261Z" level=info msg="0/1 success conditions matched"
17:45:34.261 main time="2022-06-16T17:45:34.261Z" level=info msg="Waiting for resource workflow.argoproj.io/dev-workflow-z8tr4 in namespace default resulted in retryable error: Neither success condition nor the failure condition has been matched. Retrying..."
17:45:39.270 main time="2022-06-16T17:45:39.270Z" level=info msg="Get workflows 200"
17:45:39.270 main time="2022-06-16T17:45:39.270Z" level=info msg="failure condition '{status.phase in [Error Failed]}' evaluated false"
17:45:39.270 main time="2022-06-16T17:45:39.270Z" level=info msg="success condition '{status.phase == [Succeeded]}' evaluated false"
17:45:39.270 main time="2022-06-16T17:45:39.270Z" level=info msg="0/1 success conditions matched"
```

I forward the logs from standard out to my log indexer of choice, which expects logs to be json-formatted (which most log ingestion services expect). However, I'm unable to find a way to change these "child resource creation" type logs to be json formatted. I'd love to have control over that because I want to create alerts and monitoring using the `level` property, for example.

# Use Cases

When would you use this?

When logging to a log provider, so almost always.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by Hunter-Thompson**
related to: #7361
fixes #8986

TODO: add test

Signed-off-by: Hunter-Thompson <aatman@auroville.org.in>

### CODE REVIEW

File: cmd/workflow-controller/main.go [hunk 1]
```diff
@@ -173,6 +174,7 @@ func NewRootCommand() *cobra.Command {
 	command.Flags().StringVar(&logLevel, "loglevel", "info", "Set the logging level. One of: debug|info|warn|error")
 	command.Flags().IntVar(&glogLevel, "gloglevel", 0, "Set the glog logging level")
 	command.Flags().StringVar(&logFormat, "log-format", "text", "The formatter to use for logs. One of: text|json")
+	command.Flags().StringVar(&executorLogFormat, "executor-log-format", "text", "The formatter to use for executor logs. One of: text|json")
```
Comment (terrytangyuan) [inline cmd/workflow-controller/main.go:177]: Should we reuse `log-format`?

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -506,13 +506,13 @@ func substitutePodParams(pod *apiv1.Pod, globalParams common.Parameters, tmpl *w
 
 func (woc *wfOperationCtx) newInitContainer(tmpl *wfv1.Template) apiv1.Container {
 	ctr := woc.newExecContainer(common.InitContainerName, tmpl)
-	ctr.Command = []string{"argoexec", "init", "--loglevel", getExecutorLogLevel()}
+	ctr.Command = []string{"argoexec", "init", "--loglevel", getExecutorLogLevel(), "--log-format", woc.controller.executorLogFormat()}
```
Comment (terrytangyuan) [inline workflow/controller/workflowpod.go:509]: log-format or executor-log-format?

Comment (Hunter-Thompson) [reply workflow/controller/workflowpod.go:509]: @terrytangyuan log-format.

See: https://github.com/Hunter-Thompson/argo-workflows/blob/dcd85374dc8a92100e2dea89f791da3486e48303/cmd/argoexec/commands/root.go#L39

File: cmd/workflow-controller/main.go (hunk 1 above)
Comment (Hunter-Thompson) [reply cmd/workflow-controller/main.go:177]: Reusing log-format will change it for exec, and workflow-controller.

Is this desired? I am OK with this.

Comment (Hunter-Thompson) [reply cmd/workflow-controller/main.go]: Makes sense to use `log-format`, changed it.

**Review by terrytangyuan — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**